In [2]:
import numpy as np
import pandas as pd
works = pd.DataFrame({
    '작가': ['김유정', '김유정', '김유정',
        '현진건', '현진건',
        '이상', '이상',
        '나도향', '나도향',
        '채만식', '이효석'],
    '작품': ['봄봄', '동백꽃', '만무방',
        '운수 좋은 날', 'B사감과 러브레터',
        '날개', '권태',
        '벙어리 삼룡이', '물레방아',
        '레디메이드 인생', '메밀꽃 필 무렵'],
    '연도': [1935, 1936, 1934,
        1924, 1925,
        1936, 1937,
        1925, 1925,
        1934, np.nan], # 이효석 '메밀꽃 필 무렵' 발표연도가 누락됨
    '글자수': [12500, 9800, 13600,
        11200, 7400,
        18300, 14100,
        16800, 14500,
        22000, 11800],
})
authors = pd.DataFrame({
    '작가': ['김유정', '현진건', '이상', '나도향', '채만식', '이효석', '염상섭'],
    '생년': [1908, 1900, 1910, 1902, 1902, 1907, 1897],
    '몰년': [1937, 1943, 1937, 1926, 1950, 1942, 1963],
})


In [3]:
# Q1 
# (a) - (1)
print(works.shape)
print(works.dtypes)
display(works.describe())

(11, 4)
작가         str
작품         str
연도     float64
글자수      int64
dtype: object


,연도,글자수
count,10.000000,11.000000
mean,1931.100000,13818.181818
std,5.546771,4080.641661
min,1924.000000,7400.000000
25%,1925.000000,11500.000000
50%,1934.000000,13600.000000
75%,1935.750000,15650.000000
max,1937.000000,22000.000000


(a) - (2)
(11, 4)
작가         str
작품         str
연도     float64
글자수      int64
dtype: object
연도	글자수
count	10.000000	11.000000
mean	1931.100000	13818.181818
std	5.546771	4080.641661
min	1924.000000	7400.000000
25%	1925.000000	11500.000000
50%	1934.000000	13600.000000
75%	1935.750000	15650.000000
max	1937.000000	22000.000000
(a) - (3)
shape는 DataFrame의 (행 수, 열 수)를 튜플로 반환하며, works가 11행 4열임을 확인할 수 있다. dtypes는 각 열의 자료형을 보여주는데, 연도는 NaN이 포함되어 float64, 글자수는 결측치 없이 int64로 저장된다. describe()는 숫자형 열에 대해 평균·표준편차·최솟값·최댓값 등의 요약 통계를 한번에 출력한다.

관찰: 글자수는 결측치가 없어 int64로 저장되지만, 연도 열에는 np.nan이 포함되어 있어 pandas가 NaN을 표현할 수 없는 int64 대신 float64로 처리한다.

In [4]:
# (b) - (1)
print(works.isna().sum())


작가     0
작품     0
연도     1
글자수    0
dtype: int64


(b) - (2) 
작가     0
작품     0
연도     1
글자수    0
dtype: int64
(b) - (3)


In [8]:
# (c) - (1)
works2 = works.copy()
after_1930 = works2[works2['연도'] >= 1930]
display(after_1930)

kim_or_lee = works2[works2['작가'].isin(['김유정', '이상'])]
display(kim_or_lee)

,작가,작품,연도,글자수
0,김유정,봄봄,1935.0,12500
1,김유정,동백꽃,1936.0,9800
2,김유정,만무방,1934.0,13600
5,이상,날개,1936.0,18300
6,이상,권태,1937.0,14100
9,채만식,레디메이드 인생,1934.0,22000


,작가,작품,연도,글자수
0,김유정,봄봄,1935.0,12500
1,김유정,동백꽃,1936.0,9800
2,김유정,만무방,1934.0,13600
5,이상,날개,1936.0,18300
6,이상,권태,1937.0,14100


(c) - (2)

작가	작품	연도	글자수
0	김유정	봄봄	1935.0	12500
1	김유정	동백꽃	1936.0	9800
2	김유정	만무방	1934.0	13600
5	이상	날개	1936.0	18300
6	이상	권태	1937.0	14100
9	채만식	레디메이드 인생	1934.0	22000
작가	작품	연도	글자수
0	김유정	봄봄	1935.0	12500
1	김유정	동백꽃	1936.0	9800
2	김유정	만무방	1934.0	13600
5	이상	날개	1936.0	18300
6	이상	권태	1937.0	14100

(c) - (3)
fillna로 NaN을 1936으로 대체한 뒤 astype(int)로 형변환했다. NaN이 남아 있으면 int 변환이 불가능하므로 순서가 중요하다.

In [9]:
# Q2
# (a) -  (1)
after_1930 = works2[works2['연도'] >= 1930]
display(after_1930)

kim_or_lee = works2[works2['작가'].isin(['김유정', '이상'])]
display(kim_or_lee)

,작가,작품,연도,글자수
0,김유정,봄봄,1935.0,12500
1,김유정,동백꽃,1936.0,9800
2,김유정,만무방,1934.0,13600
5,이상,날개,1936.0,18300
6,이상,권태,1937.0,14100
9,채만식,레디메이드 인생,1934.0,22000


,작가,작품,연도,글자수
0,김유정,봄봄,1935.0,12500
1,김유정,동백꽃,1936.0,9800
2,김유정,만무방,1934.0,13600
5,이상,날개,1936.0,18300
6,이상,권태,1937.0,14100


(a) - (2)
작가	작품	연도	글자수
0	김유정	봄봄	1935.0	12500
1	김유정	동백꽃	1936.0	9800
2	김유정	만무방	1934.0	13600
5	이상	날개	1936.0	18300
6	이상	권태	1937.0	14100
9	채만식	레디메이드 인생	1934.0	22000
작가	작품	연도	글자수
0	김유정	봄봄	1935.0	12500
1	김유정	동백꽃	1936.0	9800
2	김유정	만무방	1934.0	13600
5	이상	날개	1936.0	18300
6	이상	권태	1937.0	14100
(a) - (3)
불 인덱싱은 조건식이 True인 행만 선택한다. isin은 리스트 안의 값 중 하나라도 일치하면 True를 반환하므로, == 조건 두 개를 |로 연결하는 것보다 간결하게 여러 값을 필터링할 수 있다.


In [10]:
# (b)- (1)
def categorize(n: int) -> str:
    if n < 10000:
        return '짧음'
    elif n <= 15000:
        return '보통'
    else:
        return '긴'

works2['분량'] = works2['글자수'].apply(categorize)
display(works2[['작품', '글자수', '분량']])

,작품,글자수,분량
0,봄봄,12500,보통
1,동백꽃,9800,짧음
2,만무방,13600,보통
3,운수 좋은 날,11200,보통
4,B사감과 러브레터,7400,짧음
5,날개,18300,긴
6,권태,14100,보통
7,벙어리 삼룡이,16800,긴
8,물레방아,14500,보통
9,레디메이드 인생,22000,긴


(b)- (2)
	작품	글자수	분량
0	봄봄	12500	보통
1	동백꽃	9800	짧음
2	만무방	13600	보통
3	운수 좋은 날	11200	보통
4	B사감과 러브레터	7400	짧음
5	날개	18300	긴
6	권태	14100	보통
7	벙어리 삼룡이	16800	긴
8	물레방아	14500	보통
9	레디메이드 인생	22000	긴
10	메밀꽃 필 무렵	11800	보통
(b) - (3)
apply는 Series의 각 원소에 함수를 하나씩 적용한다. 직접 for문을 쓰는 것보다 pandas의 내부 최적화를 활용할 수 있고, 타입 힌트를 명시한 별도 함수를 정의함으로써 가독성과 재사용성도 높아진다.

In [11]:
# (c)- (1)
display(works2.sort_values(by=['연도', '글자수'], ascending=[True, False]))


,작가,작품,연도,글자수,분량
3,현진건,운수 좋은 날,1924.0,11200,보통
7,나도향,벙어리 삼룡이,1925.0,16800,긴
8,나도향,물레방아,1925.0,14500,보통
4,현진건,B사감과 러브레터,1925.0,7400,짧음
9,채만식,레디메이드 인생,1934.0,22000,긴
2,김유정,만무방,1934.0,13600,보통
0,김유정,봄봄,1935.0,12500,보통
5,이상,날개,1936.0,18300,긴
1,김유정,동백꽃,1936.0,9800,짧음
6,이상,권태,1937.0,14100,보통


(c) - 2
작가	작품	연도	글자수	분량
3	현진건	운수 좋은 날	1924.0	11200	보통
7	나도향	벙어리 삼룡이	1925.0	16800	긴
8	나도향	물레방아	1925.0	14500	보통
4	현진건	B사감과 러브레터	1925.0	7400	짧음
9	채만식	레디메이드 인생	1934.0	22000	긴
2	김유정	만무방	1934.0	13600	보통
0	김유정	봄봄	1935.0	12500	보통
5	이상	날개	1936.0	18300	긴
1	김유정	동백꽃	1936.0	9800	짧음
6	이상	권태	1937.0	14100	보통
10	이효석	메밀꽃 필 무렵	NaN	11800	보통
(c) - (3)
sort_values의 by와 ascending에 리스트를 넘기면 다중 기준 정렬을 한 줄로 표현할 수 있다. ascending=[True, False]는 첫 번째 키(연도)는 오름차순, 두 번째 키(글자수)는 내림차순으로 정렬함을 의미한다.

In [12]:
# Q3
# (a) - (1)
author_stats = works2.groupby('작가')['글자수'].agg(
    평균_글자수='mean',
    작품_수='count',
    총_글자수='sum'
)
display(author_stats)

,평균_글자수,작품_수,총_글자수
작가,,,
김유정,11966.666667,3,35900
나도향,15650.000000,2,31300
이상,16200.000000,2,32400
이효석,11800.000000,1,11800
채만식,22000.000000,1,22000
현진건,9300.000000,2,18600


(a) - (2)
평균_글자수	작품_수	총_글자수
작가			
김유정	11966.666667	3	35900
나도향	15650.000000	2	31300
이상	16200.000000	2	32400
이효석	11800.000000	1	11800
채만식	22000.000000	1	22000
현진건	9300.000000	2	18600
(a) - (3)
groupby 후 agg에 키워드 인수 형태로 집계 함수를 지정하면 결과 열 이름을 직접 정할 수 있어 가독성이 좋다. 한 번의 호출로 평균·개수·합계를 동시에 계산하므로 반복 연산 없이 효율적이다.

In [13]:
# (b) - (1)
print(works2['분량'].value_counts())

분량
보통    6
긴     3
짧음    2
Name: count, dtype: int64


(b) - (2)
분량
보통    6
긴     3
짧음    2
Name: count, dtype: int64
(b) - (3)
value_counts()는 Series의 각 고유값이 몇 번 등장하는지 세어 빈도 내림차순으로 반환한다. 전체 11편 중 분량 범주별로 작품이 어떻게 분포하는지 한눈에 파악할 수 있다.

In [14]:
# (c) - (1)
print(works2.groupby(['작가', '분량']).size())

작가   분량
김유정  보통    2
     짧음    1
나도향  긴     1
     보통    1
이상   긴     1
     보통    1
이효석  보통    1
채만식  긴     1
현진건  보통    1
     짧음    1
dtype: int64


(c) - (2)
작가   분량
김유정  보통    2
     짧음    1
나도향  긴     1
     보통    1
이상   긴     1
     보통    1
이효석  보통    1
채만식  긴     1
현진건  보통    1
     짧음    1
dtype: int64
(c) - (3)
groupby에 리스트로 두 열을 넘기면 두 열의 모든 조합을 기준으로 그룹을 만든다. .size()는 각 그룹의 행 수를 세어 반환하며, .agg('count')와 달리 특정 열을 지정할 필요 없이 그룹 자체의 크기를 구할 때 쓴다.

관찰: 가장 많은 작품 수를 가진 조합은 (김유정, 보통)으로 2편이며, 이는 김유정이 10,000~15,000자 안팎의 중간 분량 단편을 선호했음을 시사한다.

In [15]:
# Q4
# (a) - (1)
full = pd.merge(works2, authors, on='작가', how='left')
display(full)
print(full.shape)

,작가,작품,연도,글자수,분량,생년,몰년
0,김유정,봄봄,1935.0,12500,보통,1908,1937
1,김유정,동백꽃,1936.0,9800,짧음,1908,1937
2,김유정,만무방,1934.0,13600,보통,1908,1937
3,현진건,운수 좋은 날,1924.0,11200,보통,1900,1943
4,현진건,B사감과 러브레터,1925.0,7400,짧음,1900,1943
5,이상,날개,1936.0,18300,긴,1910,1937
6,이상,권태,1937.0,14100,보통,1910,1937
7,나도향,벙어리 삼룡이,1925.0,16800,긴,1902,1926
8,나도향,물레방아,1925.0,14500,보통,1902,1926
9,채만식,레디메이드 인생,1934.0,22000,긴,1902,1950


(11, 7)


(a)- (2)
	작가	작품	연도	글자수	분량	생년	몰년
0	김유정	봄봄	1935.0	12500	보통	1908	1937
1	김유정	동백꽃	1936.0	9800	짧음	1908	1937
2	김유정	만무방	1934.0	13600	보통	1908	1937
3	현진건	운수 좋은 날	1924.0	11200	보통	1900	1943
4	현진건	B사감과 러브레터	1925.0	7400	짧음	1900	1943
5	이상	날개	1936.0	18300	긴	1910	1937
6	이상	권태	1937.0	14100	보통	1910	1937
7	나도향	벙어리 삼룡이	1925.0	16800	긴	1902	1926
8	나도향	물레방아	1925.0	14500	보통	1902	1926
9	채만식	레디메이드 인생	1934.0	22000	긴	1902	1950
10	이효석	메밀꽃 필 무렵	NaN	11800	보통	1907	1942
(11, 7)
(a) - (3)
pd.merge는 두 DataFrame을 공통 키(on='작가')를 기준으로 합친다. how='left'이므로 왼쪽 표인 works2의 행 수(11행)가 그대로 유지되고, authors에서 해당 작가의 생년·몰년 열이 붙는다. shape를 출력해 merge 후에도 행 수가 변하지 않았음을 확인한다. 

확인 질문: authors에만 있는 '염상섭'은 full에 포함되지 않는다. how='left'는 왼쪽 표(works2)의 행만 기준으로 삼고 오른쪽에서 매칭되는 키가 있을 때만 열을 결합하므로, 오른쪽에만 존재하는 행은 결과에 나타나지 않기 때문이다.

In [16]:
# (b) - (1)
full['집필 당시 나이'] = full['연도'] - full['생년']
display(full[['작가', '작품', '연도', '생년', '집필 당시 나이']])

,작가,작품,연도,생년,집필 당시 나이
0,김유정,봄봄,1935.0,1908,27.0
1,김유정,동백꽃,1936.0,1908,28.0
2,김유정,만무방,1934.0,1908,26.0
3,현진건,운수 좋은 날,1924.0,1900,24.0
4,현진건,B사감과 러브레터,1925.0,1900,25.0
5,이상,날개,1936.0,1910,26.0
6,이상,권태,1937.0,1910,27.0
7,나도향,벙어리 삼룡이,1925.0,1902,23.0
8,나도향,물레방아,1925.0,1902,23.0
9,채만식,레디메이드 인생,1934.0,1902,32.0


(b) - (2)
작가	작품	연도	생년	집필 당시 나이
0	김유정	봄봄	1935.0	1908	27.0
1	김유정	동백꽃	1936.0	1908	28.0
2	김유정	만무방	1934.0	1908	26.0
3	현진건	운수 좋은 날	1924.0	1900	24.0
4	현진건	B사감과 러브레터	1925.0	1900	25.0
5	이상	날개	1936.0	1910	26.0
6	이상	권태	1937.0	1910	27.0
7	나도향	벙어리 삼룡이	1925.0	1902	23.0
8	나도향	물레방아	1925.0	1902	23.0
9	채만식	레디메이드 인생	1934.0	1902	32.0
10	이효석	메밀꽃 필 무렵	NaN	1907	NaN
(b) - (3)
연도 - 생년은 Series 간 뺄셈으로 행별로 자동 적용되는 벡터화 연산이다. apply나 for문 없이 한 줄로 모든 행의 나이를 동시에 계산할 수 있다.

In [17]:
# (c) - (1)
avg_age = full.groupby('작가')['집필 당시 나이'].mean().sort_values()
print(avg_age)

작가
나도향    23.0
현진건    24.5
이상     26.5
김유정    27.0
채만식    32.0
이효석     NaN
Name: 집필 당시 나이, dtype: float64


(c) - (2)
작가
나도향    23.0
현진건    24.5
이상     26.5
김유정    27.0
채만식    32.0
이효석     NaN
Name: 집필 당시 나이, dtype: float64
(c) - (3)
groupby('작가')로 작가별 그룹을 만든 뒤 ['집필 당시 나이'].mean()으로 각 작가의 평균 집필 나이를 계산한다. 뒤에 .sort_values()를 바로 이어 붙여 오름차순 정렬까지 한 줄로 처리한다.

해석: 가장 어린 나이에 작품을 발표한 작가는 이상으로, 평균 약 26~27세에 작품을 썼다. 가장 늦은 나이에 작품을 발표한 작가는 데이터상으로는 채만식(32세)이 유력하다.